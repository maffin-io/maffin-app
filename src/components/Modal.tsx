import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
