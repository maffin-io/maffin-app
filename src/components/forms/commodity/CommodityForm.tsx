import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { SingleValue } from 'react-select';

import { Commodity } from '@/book/entities';
import Selector from '@/components/selectors/Selector';

const resolver = classValidatorResolver(Commodity, { validator: { stopAtFirstError: true } });

export type FormValues = {
  namespace: string,
  mnemonic: string,
};

export type CommodityFormProps = {
  onSave: Function,
};

const BASE_CURRENCIES = [
  { label: 'EUR' },
  { label: 'USD' },
  { label: 'SGD' },
  { label: 'GBP' },
];

export default function CommodityForm({ onSave }: CommodityFormProps): JSX.Element {
  const form = useForm<FormValues>({
    mode: 'onChange',
    resolver,
  });

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, onSave))}>
      <fieldset className="text-sm my-5">
        <Controller
          control={form.control}
          name="mnemonic"
          render={({ field, fieldState }) => (
            <>
              <Selector<{ label: string }>
                id="mnemonicInput"
                labelAttribute="label"
                options={BASE_CURRENCIES}
                onChange={(newValue: SingleValue<{ label: string }> | null) => {
                  field.onChange(newValue?.label);
                }}
                placeholder="Choose your currency"
                isClearable={false}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <input
        {...form.register(
          'namespace',
        )}
        hidden
        value="CURRENCY"
        type="text"
      />

      <div className="flex w-full justify-center">
        <button className="btn btn-primary" type="submit" disabled={!form.formState.isValid}>
          Save
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, onSave: Function) {
  const mainCommodity = await Commodity.create({ ...data }).save();
  await onSave(mainCommodity);
}
