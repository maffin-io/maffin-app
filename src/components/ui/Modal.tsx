import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
