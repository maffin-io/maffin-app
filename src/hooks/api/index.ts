import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
