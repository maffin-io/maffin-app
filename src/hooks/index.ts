export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
