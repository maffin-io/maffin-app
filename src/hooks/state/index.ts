export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
