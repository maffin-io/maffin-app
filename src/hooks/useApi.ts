import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
