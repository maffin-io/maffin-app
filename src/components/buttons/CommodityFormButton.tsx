import React from 'react';
import Modal from 'react-modal';
import { BiPlusCircle } from 'react-icons/bi';

import { DataSourceContext } from '@/hooks';
import CommodityForm from '@/components/forms/commodity/CommodityForm';
import type { Commodity } from '@/book/entities';

export interface AccountFormButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  action?: 'add' | 'update' | 'delete',
  defaultValues?: Partial<Commodity>,
}

export default function AccountFormButton({
  action = 'add',
  defaultValues,
  children,
  className = 'btn btn-primary',
  ...props
}: AccountFormButtonProps): JSX.Element {
  const { save } = React.useContext(DataSourceContext);
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);

  let title = 'Add commodity';
  if (action === 'update') {
    title = `Edit ${defaultValues?.mnemonic} commodity`;
  }
  if (action === 'delete') {
    title = 'Confirm you want to remove this commodity';
  }

  return (
    <>
      {/* @ts-ignore */}
      <Modal
        isOpen={isModalOpen}
        overlayClassName="overlay"
        className="modal"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setIsModalOpen(false)}
        >
          X
        </button>
        <span>{title}</span>
        <CommodityForm
          action={action}
          defaultValues={defaultValues}
          onSave={() => {
            save();
            setIsModalOpen(false);
          }}
        />
      </Modal>
      <button
        id="add-commodity"
        type="button"
        onClick={() => setIsModalOpen(!isModalOpen)}
        className={className}
        {...props}
      >
        {
          children
          || (
            <>
              <BiPlusCircle className="mr-1" />
              Add Commodity
            </>
          )
        }
      </button>
    </>
  );
}
