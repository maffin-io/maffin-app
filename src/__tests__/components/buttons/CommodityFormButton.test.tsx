import React from 'react';
import {
  act,
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import Modal from 'react-modal';

import CommodityFormButton from '@/components/buttons/CommodityFormButton';
import CommodityForm from '@/components/forms/commodity/CommodityForm';
import { DataSourceContext } from '@/hooks';
import type { DataSourceContextType } from '@/hooks';

jest.mock('react-modal', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="Modal">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/commodity/CommodityForm', () => jest.fn(
  () => <div data-testid="CommodityForm" />,
).mockName('CommodityForm'));

describe('CommodityFormButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders hidden modal on mount', async () => {
    render(<CommodityFormButton />);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(CommodityForm).toHaveBeenLastCalledWith(
      {
        action: 'add',
        onSave: expect.any(Function),
        defaultValues: undefined,
      },
      {},
    );
  });

  it('renders hidden modal on mount with update', async () => {
    render(<CommodityFormButton action="update" />);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(CommodityForm).toHaveBeenLastCalledWith(
      {
        action: 'update',
        onSave: expect.any(Function),
        defaultValues: undefined,
      },
      {},
    );
  });

  it('renders hidden modal on mount with delete', async () => {
    render(<CommodityFormButton action="delete" />);
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(CommodityForm).toHaveBeenLastCalledWith(
      {
        action: 'delete',
        onSave: expect.any(Function),
        defaultValues: undefined,
      },
      {},
    );
  });

  it('opens modal when clicking the button', async () => {
    render(
      <CommodityFormButton />,
    );

    const button = await screen.findByRole('button', { name: /add commodity/i });
    fireEvent.click(button);

    const modal = await screen.findByTestId('Modal');
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );
    expect(modal).toMatchSnapshot();
  });

  it('closes modal when clicking the X button', async () => {
    render(
      <CommodityFormButton />,
    );

    const button = await screen.findByRole('button', { name: /add commodity/i });
    fireEvent.click(button);

    await screen.findByTestId('Modal');
    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );

    const xButton = screen.getByRole('button', { name: 'X' });
    fireEvent.click(xButton);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });

  it('on form save closes modal and saves', async () => {
    const mockSave = jest.fn();
    render(
      <DataSourceContext.Provider value={{ save: mockSave as Function } as DataSourceContextType}>
        <CommodityFormButton />
      </DataSourceContext.Provider>,
    );

    // open modal to prove that onSave closes it
    const button = await screen.findByRole('button', { name: /add commodity/i });
    fireEvent.click(button);

    const { onSave } = (CommodityForm as jest.Mock).mock.calls[0][0];
    act(() => onSave());
    expect(mockSave).toBeCalledTimes(1);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });

  it('passes values to CommodityForm', async () => {
    const mockSave = jest.fn();
    render(
      <DataSourceContext.Provider value={{ save: mockSave as Function } as DataSourceContextType}>
        <CommodityFormButton
          defaultValues={{
            mnemonic: 'EUR',
          }}
        />
      </DataSourceContext.Provider>,
    );

    // open modal to prove that onSave closes it
    const button = await screen.findByRole('button', { name: /add commodity/i });
    fireEvent.click(button);

    expect(CommodityForm as jest.Mock).toBeCalledWith(
      {
        action: 'add',
        onSave: expect.any(Function),
        defaultValues: {
          mnemonic: 'EUR',
        },
      },
      {},
    );
  });

  it('renders the children', async () => {
    render(
      <CommodityFormButton>
        <span>hello</span>
      </CommodityFormButton>,
    );

    await screen.findByText('hello');
  });
});
