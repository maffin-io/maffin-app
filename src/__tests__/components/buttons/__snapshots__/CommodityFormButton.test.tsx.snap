// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommodityFormButton opens modal when clicking the button 1`] = `
<div
  data-testid="Modal"
>
  <button
    class="float-right"
    type="button"
  >
    X
  </button>
  <span>
    Add commodity
  </span>
  <div
    data-testid="CommodityForm"
  />
</div>
`;
